src/persistQueryClient/PersistQueryClientProvider.tsx

```diff
@@ -0,0 +1,46 @@
+import React from 'react'
+
+import { persistQueryClient, PersistQueryClientOptions } from './persist'
+import { QueryClientProvider, QueryClientProviderProps } from '../reactjs'
+
+export interface PersistQueryClientProviderProps
+  extends QueryClientProviderProps {
+  persistOptions: Omit<PersistQueryClientOptions, 'queryClient'>
+  loading?: React.ReactNode
+}
+
+export const PersistQueryClientProvider = ({
+  client,
+  loading,
+  children,
+  persistOptions,
+  ...props
+}: PersistQueryClientProviderProps): JSX.Element => {
+  const [initialized, setInitialized] = React.useState(false)
+  const options = React.useRef(persistOptions)
+  React.useEffect(() => {
+    let unsubscribe: (() => void) | undefined
+
+    async function run() {
+      unsubscribe = await persistQueryClient({
+        ...options.current,
+        queryClient: client,
+      })
+      setInitialized(true)
+    }
+
+    void run()
+
+    return unsubscribe
+  }, [client])
+
+  if (!initialized) {
+    return <>{loading}</>
+  }
```

Comment:
I guess this wouldn't work with SSR? I mean it would work, but you would only get `loading` rendered on the server. Is this even a use-case: hydrating from the server and restoring from local storage? I fear that if we make an exception for SSR here, we will get hydration mismatches. Also rendering on the server and then overriding that with state from localStorage seems weird... what do you think?

@Ephem fyi

Reply:
Hmm, I haven't thought about the persisting use case much, but yeah, does seem a little weird. I guess one use case could be if you have a very long cache time in a CDN or something (or static generation) where the data in the server rendered html can be pretty old. If the user has visited the site before and persisted fresher data, I _guess_ it could be nice if you first hydrated the server rendered html and as soon as the app has hydrated, you update the data to the fresher version that was previously persisted.

Feel like a lot of hoops to jump through though, if data freshness matters you probably wanna revalidate by refetching anyway.

Reply:
The problem is that you can't really use this approach together with SSR, as its an async rehydrate. What happens if you use the two together, but don't have something in localstorage?

- SSR will render `<>{loading}</>`
- on the client, we'll render `<>{loading}</>`, followed by a read from localstorage
- we don't read anything, so we just client render the whole page

That basically means if you use this component, you can't SSR anything on the page, which seems bad :/

redux-persist has uses the same approach with the [PersistGate](https://github.com/rt2zz/redux-persist/blob/d8b01a085e3679db43503a3858e8d4759d6f22fa/src/integration/react.ts#L16) component, but they also have a ton of open issues regarding SSR, e.g.:

- https://github.com/rt2zz/redux-persist/issues/1008
- https://github.com/vercel/next.js/issues/8240#issuecomment-647699316

all the provided solutions are basically: don't use `PersistGate`, or use the children render function, which basically also doesn't render the Gate, which means that the restore actions might be dispatched _after_ the app has started rendering, which is the exact same issue I was trying to solve.

if we disable the Gate on the server only, as [suggested on stackoverflow](https://stackoverflow.com/questions/58238525/gatsby-not-generating-correct-static-html-files/62501513#62501513), we will get hydration mismatches, because the server renderes the whole markup while the client only renders the loading gate.

I don't see a solution right now. Maybe we should just document this properly, that this doesn't work with SSR and that people can still subscribe to / restore from their storage whenever they want manually with the provided functions?

Reply:
Ahh, yes, now that I actually read the code (😬 ) I see that this components job is specifically to delay rendering until this is done, which is as you say incompatible with SSR at a fundamental level. 😄 

If there exists a low level API to restore a persist manually, that should be enough for any wonky SSR usecases people might have.

Reply:
@Ephem please have a look here: 5c4de915

As briefly discussed on Discord, I added another `IsHydrating` context and I've deferred observer subscriptions until we have finished hydrating.

Further, if we are hydrating, I've put the query into the new `fetchState: paused`. We could also provide a `pausedReason` or something similar, but I think at least it gives the users the opportunity to _not_ show a loading spinner, because the `status` of the query will be `loading` in the meantime. However, no actual request is being made. We could also put it to `fetchStatus: 'idle'` if we want.

Now currently, this only applies to the new `PersistQueryClientProvider`, but maybe the new hydrate that should run in a useEffect could use the same mechanism? Or would it need to be a separate context then? If we do SSR + persistence, we would somehow need to make sure that restoring from localStorage would run _after_ hydration from SSR is finished 🤔

Reply:
So it's still in a loading state (`isLoading` is true), but you could look at the `fetchStatus` specifically to avoid showing a spinner if you want to take it a step further? I like it!

For SSR hydration/incremental page hydration this will get trickier. On the initial hydrate from the server markup, `isHydrating` should be false because we already have the data in the cache (hydration happened on cache/queryClient creation so to speak). Only _after_ that should we set `isHydrating`.

I think it would be nice to reuse the same context for all hydration if we can, but I'm also not sure about how the SSR+persistance use case should work, it's tricky.

Just to think out loud, I guess this provider as well as the planned hydration provider are both really "gates" so to speak. While the hooks versions of these would work fine on their own, the real problem with those is that the queries would still fire and trigger a fetch, even if we are currently in the process of hydrating "something". (We aren't really sure what we are hydrating is everything that is actually needed to render btw, so by blocking all queries some we actually do need to fetch might get slightly delayed, but I think we are talking about a neglectable amount of time compared to network latency.)

If we are rendering the `PersistQueryClientProvider` on the server, I guess we could want one of three things to happen:

1. We are telling the server we'll just go ahead and restore on the client, so don't worry about fetching any data, just render loading state placeholders
2. We are telling the server "on the client, I'm going to try to restore fresh data from some external cache, but since you don't have that available now you'll still need to go ahead and fetch all data needed"
3. Throw an error

Nr 1 just seems pointless to me. Nr 2 seems pretty complex since now the initial state on the client _can't_ be `isLoading && fetchStatus: 'paused'` since that's not the status it had on the server, there it was successful. We would have to delay restoring until after SSR hydration is complete. But if we have data from the server, we still don't want the queries to be in a `isLoading` state, since we already have data to show.

Or maybe this is even more complex? If we first hydrate from the server and then there are some components that were client rendered that are still waiting to fetch data, then in the persistance case _those_ queries should now be paused, and the queries that now have data should be successful, but in a refetching state while restoring..?

🥴

Reply:
> So it's still in a loading state (isLoading is true), but you could look at the fetchStatus specifically to avoid showing a spinner if you want to take it a step further? I like it!

Yes, this is a new feature in v4. Currently, we use `paused` only when a query is paused due to no internet connection, but we could widen that case. We currently have `fetchStatus: 'idle' | 'paused' | 'fetching'`. we could even have a fourth state but I don't see the need for that. I just [blogged](https://tkdodo.eu/blog/offline-react-query) about it yesterday :)

> For SSR hydration/incremental page hydration this will get trickier. On the initial hydrate from the server markup, isHydrating should be false because we already have the data in the cache (hydration happened on cache/queryClient creation so to speak). Only after that should we set isHydrating (on page navigation and the like).

The plan is to have initial hydration happen via the `initialState` we pass to the `QueryClientProvider`, right? This would by synchronous and we wouldn't need any isHydrating state 👍 

> Just to think out loud, I guess this provider as well as the planned hydration provider are both really "gates" so to speak. While the hooks versions of these would work fine on their own, the real problem with those is that the queries would still fire and trigger a fetch, even if we are currently in the process of hydrating "something". (We aren't really sure what we are hydrating is everything that is actually needed to render btw, so by blocking all queries some we actually do need to fetch might get slightly delayed, but I think we are talking about a neglectable amount of time compared to network latency.)

Yes, I think that's correct. The `gates` basically aim to stop triggering a fetch. The observer is created, but not subscribed, so no fetch is initiated. That also means that `initialData` is taken into account, so without `initialDataUpdatedAt`, that data would be newer than the restored one.
Anyhow, the delay is acceptable I guess. You'd explicitly _want_ to delay fetching while restoring, because if you have _staleTime_ set and you hydrate sufficiently fresh data, there shouldn't be any data fetching at all!

> Or maybe this is even more complex? If we first hydrate from the server and then there are some components that were client rendered that are still waiting to fetch data, then in the persistance case those queries should now be paused, and the queries that now have data should be successful, but in a refetching state while restoring..?

For the combo of the two (SSR + persist reloading), I was think more of the SSG use-case:

- you pre-render a site on the server, it produces markup, we send that to the client.
- the client hydrates that via `initialState`, synchronously.
- then, the `persist` provider would try to restore additional state from the localStorage. This might be queries that we haven't SSGed on the server, or it might be newer variants of already cached queries. On the first visit, nothing would happen because you have no localStorage data. But if you do, you would get your data restored because it very likely has a newer `dataUpdatedAt` timestamp.

With SSR, this would likely have the opposite ordering: Since the SSRed data has a brand new timestamp, it is likely newer than your localStorage data, so the persist plugin would read out data but _not_ hydrate it. I guess that's okay and needs to be documented.

And for additional page navigations where we aim to hydrate in useEffect, the persist "gate" wouldn't need to do anything because restoring should only do something on page load - so we shouldn't have any conflicts here...

Reply:
I just read this again and I think a bunch of these issues are more straightforward if normal hydration is sync via putting the data on a context like we've talked about.

Just wanted to note on the SSG/SSR usecase, since I never replied to that, that yes, often what you are describing is the case, but if you cache SSR you end up with the same case as the SSG one you are describing. 😄